x-pack/plugins/fleet/server/services/metrics/fetch_agent_metrics.ts

```diff
@@ -0,0 +1,260 @@
+      appContextService.getLogger().debug('Index .fleet-agents does not exist yet.');
+    } else {
+      throw error;
+    }
+    return [];
+  }
+};
+
+export const getUpgradingSteps = async (
+  esClient: ElasticsearchClient,
+  abortController: AbortController
+): Promise<UpgradingSteps> => {
+  const upgradingSteps = {
+    requested: 0,
+    scheduled: 0,
+    downloading: 0,
+    extracting: 0,
+    replacing: 0,
+    restarting: 0,
+    watching: 0,
+    rollback: 0,
+    failed: 0,
+  };
+  try {
+    const response = await esClient.search(
+      {
+        index: AGENTS_INDEX,
+        query: {
+          bool: {
+            filter: [
+              {
+                term: {
+                  active: 'true',
+                },
+              },
+            ],
+          },
+        },
+        size: 0,
+        aggs: {
+          upgrade_details: {
+            terms: { field: 'upgrade_details.state' },
+          },
+        },
+      },
+      { signal: abortController.signal }
+    );
+    ((response?.aggregations?.upgrade_details as any).buckets ?? []).forEach((bucket: any) => {
+      switch (bucket.key) {
+        case 'UPG_REQUESTED':
+          upgradingSteps.requested = bucket.doc_count;
+          break;
+        case 'UPG_SCHEDULED':
+          upgradingSteps.scheduled = bucket.doc_count;
+          break;
+        case 'UPG_DOWNLOADING':
+          upgradingSteps.downloading = bucket.doc_count;
+          break;
+        case 'UPG_EXTRACTING':
+          upgradingSteps.extracting = bucket.doc_count;
+          break;
+        case 'UPG_REPLACING':
+          upgradingSteps.replacing = bucket.doc_count;
+          break;
+        case 'UPG_RESTARTING':
+          upgradingSteps.restarting = bucket.doc_count;
+          break;
+        case 'UPG_WATCHING':
+          upgradingSteps.watching = bucket.doc_count;
+          break;
+        case 'UPG_ROLLBACK':
+          upgradingSteps.rollback = bucket.doc_count;
+          break;
+        case 'UPG_FAILED':
+          upgradingSteps.failed = bucket.doc_count;
+          break;
+        default:
+          break;
+      }
+    });
+    return upgradingSteps;
+  } catch (error) {
+    if (error.statusCode === 404) {
+      appContextService.getLogger().debug('Index .fleet-agents does not exist yet.');
+    } else {
+      throw error;
+    }
+    return upgradingSteps;
+  }
+};
+
+export const getUnhealthyReason = async (
+  esClient: ElasticsearchClient,
+  abortController: AbortController
+): Promise<UnhealthyReason> => {
+  const unhealthyReason = {
+    input: 0,
+    output: 0,
+    other: 0,
+  };
+  try {
+    const response = await esClient.search(
+      {
+        index: AGENTS_INDEX,
+        query: {
+          bool: {
+            filter: [
+              {
+                term: {
+                  active: 'true',
+                },
+              },
+              {
+                terms: {
+                  last_checkin_status: ['error', 'degraded'],
```

Comment:
querying agent docs that have `last_checkin_status`: `error` or `degraded`, assuming those are unhealthy with `components` info